File: py/src/util/resources.py
from src.constants import RESOURCES_CATEGORIES, RESOURCES_DIR, RESUME_MD_PATH


def load_resource(name: str) -> str:
    if name == "resume":
        path = RESUME_MD_PATH
    else:
        path = RESOURCES_DIR / f"{name}.md"

    if not path.exists():
        return f"Resource '{name}' not found. Please create {path}"

    return path.read_text()


def list_resources() -> list[str]:
    resources = []
    if RESUME_MD_PATH.exists():
        resources.append("resume")
    for expected in RESOURCES_CATEGORIES:
        if (RESOURCES_DIR / f"{expected}.md").exists():
            resources.append(expected)

    return resources


def search_resources(query: str) -> dict[str, list[str]]:
    results = {}
    query_lower = query.lower()
    for resource_name in list_resources():
        content = load_resource(resource_name)
        matching_lines = [line for line in content.splitlines() if query_lower in line.lower()]

        if matching_lines:
            results[resource_name] = matching_lines

    return results

File: tsx/src/util/resources.ts
import { existsSync } from "fs";
import { resolve } from "path";
import {
  RESOURCES_CATEGORIES,
  RESOURCES_DIR,
  RESUME_MD_PATH,
} from "../constants.js";

export async function loadResource(name: string): Promise<string> {
  const path =
    name === "resume" ? RESUME_MD_PATH : resolve(RESOURCES_DIR, `${name}.md`);

  const file = Bun.file(path);
  if (!(await file.exists())) {
    return `Resource '${name}' not found. Please create ${path}`;
  }

  return await file.text();
}

export function listResources(): string[] {
  const resources: string[] = [];

  if (existsSync(RESUME_MD_PATH)) {
    resources.push("resume");
  }

  for (const category of RESOURCES_CATEGORIES) {
    const path = resolve(RESOURCES_DIR, `${category}.md`);
    if (existsSync(path)) {
      resources.push(category);
    }
  }

  return resources;
}

export async function searchResources(
  query: string,
): Promise<Record<string, string[]>> {
  const results: Record<string, string[]> = {};
  const queryLower = query.toLowerCase();

  for (const resourceName of listResources()) {
    const content = await loadResource(resourceName);
    const matchingLines = content
      .split("\n")
      .filter((line) => line.toLowerCase().includes(queryLower));

    if (matchingLines.length > 0) {
      results[resourceName] = matchingLines;
    }
  }

  return results;
}
